src/core/structure/sheet.ts

```diff
@@ -396,18 +435,114 @@ export default class Sheet {
     return this.cellData.get(cellKey)!;
   }
 
-  private resolveCell(cell: Cell): boolean {
-    const value = this.expressionHandler.evaluate(cell.formula);
-    if (value == null) {
-      cell.state = CellState.INVALID_EXPRESSION;
-      return false;
+  /**
+   * Returns true if the value of the cell has changed.
+   */
+  private resolveCell(cell: Cell, colKey: ColumnKey, rowKey: RowKey): boolean {
+    const evalResult = this.expressionHandler.evaluate(cell.formula);
+    if (!evalResult) {
+      const prevState = cell.state;
+      cell.setState(CellState.INVALID_EXPRESSION);
+      return prevState == CellState.OK; // Consider the cell's value changed if its state changes from OK to invalid.
     }
```

Comment:
> Consider the cell's value changed if its state changes from OK to invalid.
I think this consideration doesn't hold true if we had a formula bar. For this statement, if I input an invalid statement then I edit the statement but it is also invalid it should still counts as a change, because we changed the formula, yeah the value hasn't changed but the formula did.

Reply:
Yes, but this is in `resolveCell`. We're only concerned with the value there, since that's what `resolveCell` is setting. You can see in `setCell` that an event is emitted regardless of the result. The return value is used when updating references, in which case it only matters if the cell's value changes, not the formula.